File: Data Structures & Algorithms/max-area-of-island/submission-3.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    maxAreaOfIsland(grid) {
        const m = grid.length;
        const n = grid[0].length;
        let size = 0;

        function dfs(r, c) {
            if(r < 0 || r >= m || c < 0 || c >= n || grid[r][c] == 0) return;

            grid[r][c] = 0;
            size++;

            dfs(r - 1, c);
            dfs(r + 1, c);
            dfs(r, c - 1);
            dfs(r, c + 1);
        }

        let ans = 0;
        for(let i = 0; i < m; ++i) {
            for(let j = 0; j < n; ++j) {
                if(grid[i][j] == 1) {
                    size = 0;
                    dfs(i, j);
                    ans = Math.max(ans, size);
                }
            }
        }
        return ans;
    }
}

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-3.java
class Solution {
    private int[][] dp;
    private int[][] dirs = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
    public int longestIncreasingPath(int[][] matrix) {
        if(matrix == null || matrix.length == 0 || matrix[0].length == 0) return 0;
        int m = matrix.length, n = matrix[0].length;
        dp = new int[m][n];

        int ans = 0;
        for(int i = 0; i < m; ++i) {
            for(int j = 0; j < n; ++j) {
                ans = Math.max(ans, dfs(matrix, i, j, m, n));
            }
        }
        return ans;
    }
    private int dfs(int[][] matrix, int r, int c, int m, int n) {
        if(dp[r][c] != 0) return dp[r][c];

        int maxLen = 1;
        for(int[] d : dirs) {
            int nr = r + d[0], nc = c + d[1];
            if(nr >= 0 && nc >= 0 && nr < m && nc < n && matrix[nr][nc] < matrix[r][c]) {
                maxLen = Math.max(maxLen, 1 + dfs(matrix, nr, nc, m, n));
            }
        }
        return dp[r][c] = maxLen;
    } 
}
